src/cli/field_data.rs

```diff
@@ -0,0 +1,27 @@
+use anyhow::Result;
+use serde::{Deserialize, Serialize};
+
+use lurk::field::{LanguageField, LurkField};
+
+/// A wrapper for data whose deserialization depends on a certain LurkField
+#[derive(Serialize, Deserialize)]
+pub struct FieldData {
+    pub field: LanguageField,
+    data: Vec<u8>,
+}
+
+#[allow(dead_code)]
+impl FieldData {
```

Comment:
Ah, but with this version you are moving the conversion to/from bytes at the wrapping time. I suspect you could peg what you want on a 
```rust
struct Labeled<T: Serialize + DeserializeOwned> {
  label: Language field,
  val: T,
}
```

Reply:
For this case I do intend to deserialize in two steps. We first read from FS to know the field and then we read from the vector to get the data with the correct field elements

Reply:
Yes. You can definitely have those two steps, but in sequence within the same function, with the above structure.

Reply:
I don't get the idea. I don't want to deserialize the vector of bytes if there is some inconsistency in the field. I want to error earlier. In other words, the vector is desirable

Reply:
For the record, I did try something like you propose in my first attempt but then I got stuck because Rust doesn't have dependent types. I wanted `T<F>` where `F: LurkField`

Reply:
I'm not sure what you're getting at about dependent types, but can you accomplish your goal with a marker trait? Here's some code that at least compiles, in case it gives you any ideas. I didn't try to write everything, so what I did might not be addressing what you care about, but maybe it helps.

```rust
use std::marker::PhantomData;

pub trait SLurkField: LurkField + Serialize {}
pub trait UsesField<F: SLurkField>: Serialize {}

#[derive(Serialize, Deserialize)]
pub struct Labeled<F: SLurkField, T: UsesField<F>> {
    field: LanguageField,
    val: T,
    _p: PhantomData<F>,
}

impl<F: SLurkField, T: UsesField<F>> From<T> for Labeled<F, T> {
    fn from(x: T) -> Self {
        Self {
            field: F::FIELD,
            val: x,
            _p: Default::default(),
        }
    }
}

#[derive(Serialize)]
struct MyStruct<F: SLurkField> {
    data: F,
}

impl<F: SLurkField> UsesField<F> for MyStruct<F> {}

#[allow(dead_code)]
impl<F: SLurkField, T: UsesField<F>> Labeled<F, T> {
    pub fn serialize(&self) -> Result<Vec<u8>> {
        Ok(bincode::serialize(&self.val)?)
    }
}

fn this_compiles<F: SLurkField>() -> Result<Vec<u8>> {
    let m = MyStruct { data: F::ZERO };
    let labeled: Labeled<_, _> = m.into();

    labeled.serialize()
}
```

Reply:
Sorry, let me try to explain better.
Actually, I need `FieldData` to only know about `LanguageField` and not be parametrized on `LurkField`.

When I read `FieldData` from the file system, I haven't deserialized the original data yet. Before I do that, I need to know to which field I might be deserializing to. This is why `FieldData::extract` is parametrized on a field:

```rust
pub fn extract<'a, F: LurkField, T: Deserialize<'a>>(&'a self) -> Result<T> {
    if self.field != F::FIELD {
        bail!("Invalid field: {}. Expected {}", &self.field, &F::FIELD)
    }
    Ok(bincode::deserialize(&self.data)?)
}
```
This is a 2-step wrapper which assures the consistency of the deserialization.

Reply:
Hmmm.

As partially unrelated topic, I don't think any strategy that persists important Lurk data using an ad-hoc/bincode serialization of `LanguageField` is a good idea. It would be A Bad Thing if changes to that enum (which are so likely they are predictable, so predictable we should plan for them) led to supposedly durable data becoming unreadable.

Reply:
How can we encode that information then? Create a duplicated struct that we try to assure its stability?

Reply:
That works for commitments, but `lurk verify <proof_id>` would need to ask the user for the field, which is annoying. So I thought that we might as well just use the same infra that's already available to simplify the directory structure.

But I will drop the use of `FieldData` for commitments because we always know the field we're operating at when we open commitments

Reply:
Aren't proofs also always with respect to a known prover/verifier, which is itself necessarily parameterized on curve+field?

Also, why the heck are proofs (apparently) being stored with an id that is the *timestamp* rather than using content-addressing as previously?

---

Question: Are you trying to support verifying *any* Lurk proof, or do we assume that a given REPL session will only verify proofs of statements in the current field?

Either way, if you content-address the proofs, you should be able to use the symlink approach as above. (NOTE: in that model, you will need to check the actual location of resolved symlinks to get that meta-data — but I still think that's not what should be needed here.)

---
Opinion: Proof IDs should be the content address of the *claim* — just as they are in `fcomm` and `clutch`. This is actually *important* because it allows for caching of proofs. It's easy to imagine applications for which equivalent proofs are requested more than once (even many, many times). For example, that's how the current Lurk website works (or was designed to): we serve real proofs of expected claims in a way that is cost effective but still accurate.

A real outsourced-but-provable computation service could do the same.

Reply:
Going back to the core point: https://gist.github.com/huitseeker/2c6bc8bbecba59cdd4e8c6cbfadebc5c